kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractExactlyOnceKafkaOutputOperator.java

```diff
@@ -0,0 +1,366 @@
+ * @category Messaging
+ * @tags output operator
+ *
+ * @since 3.5
+ */
+@org.apache.hadoop.classification.InterfaceStability.Evolving
+public abstract class AbstractExactlyOnceKafkaOutputOperator<T> extends AbstractKafkaOutputOperator<String, T>
+    implements Operator.CheckpointNotificationListener
+{
+  private transient String key;
+  private transient String appId;
+  private transient Integer operatorId;
+  private transient Long windowId;
+  private transient Map<T, Integer> partialWindowTuples = new HashMap<>();
+  private transient KafkaConsumer consumer;
+
+  private WindowDataManager windowDataManager = new FSWindowDataManager();
+  private final int KAFKA_CONNECT_ATTEMPT = 10;
+  private final String KEY_SEPARATOR = "#";
+  private final String KEY_SERIALIZER = "org.apache.kafka.common.serialization.StringDeserializer";
+  private final String VALUE_SERIALIZER = "org.apache.kafka.common.serialization.StringDeserializer";
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+
+    this.operatorId = context.getId();
+    this.windowDataManager.setup(context);
+    this.appId = context.getValue(Context.DAGContext.APPLICATION_ID);
+    this.key = appId + KEY_SEPARATOR + (new Integer(operatorId));
+    this.consumer = KafkaConsumerInit();
+
+    setProperty(ACKS_CONFIG, "all");
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    this.windowId = windowId;
+
+    if (windowId == windowDataManager.getLargestRecoveryWindow()) {
+      rebuildPartialWindow();
+    }
+  }
+
+  @Override
+  public void checkpointed(long windowId)
+  {
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+    try {
+      windowDataManager.deleteUpTo(operatorId, windowId);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void beforeCheckpoint(long windowId)
+  {
+  }
+
+  @Override
+  public void teardown()
+  {
+    consumer.close();
+    super.teardown();
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (!partialWindowTuples.isEmpty()) {
+      // Throw exception
+    }
+
+    try {
+      this.windowDataManager.save(getPartitionsAndOffsets(), operatorId, windowId);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public WindowDataManager getWindowDataManager()
+  {
+    return windowDataManager;
+  }
+
+  public void setWindowDataManager(WindowDataManager windowDataManager)
+  {
+    this.windowDataManager = windowDataManager;
+  }
+
+  private boolean doesKeyBelongsToThisInstance(Integer operatorId, String key)
+  {
+    String[] split = key.split(KEY_SEPARATOR);
+
+    if (split.length != 2) {
+      return false;
+    }
+
+    if ((Integer.parseInt(split[1]) == operatorId) && (split[0].equals(appId))) {
```

Comment:
why not have a partitionId instead of appId.A re-launch will give a new applicationId so the second check will not work.